tests/specs/lemmas.k

```diff
@@ -291,6 +297,12 @@ module LEMMAS-HASKELL [symbolic, kore]
     rule 0 <=Int #sizeByteArray ( _ )     => true [smt-lemma]
     rule 0 <=Int lengthBytes ( _ )        => true [smt-lemma]
 
-    rule bool2Word( B:Bool ) ==Int 1 => B         [simplification]
-    rule bool2Word( B:Bool ) ==Int 0 => notBool B [simplification]
+    //Simplification of bool2Word() ==Int CONCRETE, #buf() ==K CONCRETE
+    rule I                   ==Int bool2Word( B:Bool ) => bool2Word(B) ==Int I  [simplification, concrete(I)]
+    rule bool2Word( B:Bool ) ==Int I                   => B ==K word2Bool(I)    [simplification, concrete(I)]
+
+    rule BA:ByteArray     ==K #buf( 32, DATA ) => #buf( 32, DATA ) ==K BA       [simplification, concrete(BA)]
+    rule #buf( 32, DATA ) ==K BA:ByteArray     => DATA ==Int #asInteger(BA)
+      requires #sizeByteArray(BA) <=Int 32                                      [simplification, concrete(BA)]
+
```

Comment:
I think this is the nontrivial change in this PR.

@denis-bogdanas Could you please explain why this is needed? I mean, why the previous lemmas that you are going to delete above are not enough?

Reply:
Also, could you explain the scenario when the equality `#buf() ==K CONCRETE` occurs?

Reply:
1. Because tests in `lemmas-spec.k` are run on both Java and Haskell.
2. To increase compatibility between the 2 sets of lemmas.

`#buf` equality occurs in symbolic testing, in `assume()` custom function. It could have been written differently that wouldn't trigger this term but chances are we'll need this lemma in other places. See `lemmas-spec.k` for real-life tests.

Reply:
I cannot see why lemmas-spec.k is relevant for bool2Word lemmas.

Reply:
Also, do you think that it is better to have the lemmas specific to the symbolic testing in this common lemma file?

Reply:
@daejunpark I think we should put them in the common lemmas file unless (i) we have a specific reason not to, or (ii) someone shows an example where it actually hurts performance.

We should not have a huge fragmented set of lemmas spread out everywhere because we're afraid it _might_ hurt performance. If you have concrete evidence that it _does_ hurt performance somewhere, we need to make a test for it.

Reply:
1. I'm not sure what you mean. There are tests for `bool2Word` lemmas in lemmas-spec.k.
2. I think it's better to have the setup for Haskell and Java as close as possible. When there is no big effort doing that. This will help porting stuff from one backend to another, and who knows maybe comparing deductive power of the 2 backend. If maintenance is a concern then we should add comments about why each group of lemmas was needed. Or better, for each test, if it was inspired from real life stuck configurations, write where it was taken from.

3. Regarding performance. In Haskell regular rules and builtins take priority, so lemmas will be tried only if regular rules couldn't apply. In Java Builtins apply first then all the rest. In addition function evaluations are cached so performance drawback is minimal.

Generally lemmas hugely affect performance when:
- they match very general terms, especially terms in side conditions of many rules, like `A <=Int MaxUInt256`
- they have side conditions
- side conditions don't match. Triggering attempts to apply lemmas over and over again.

Terms above are not that common.

Reply:
@daejunpark to clarify, we have a spec called `tests/specs/functional/lemmas-spec.k` which tests the functional simplifications that show up in the `lemmas.k` file. This test is run on both Java and Haskell backend, and is the place to put regressions in terms of simplification power that occur (to ensure that the same terms will remain simplifiable in the future). @denis-bogdanas has provided some tests of his changes there (at my request), meaning they need to be simplifiable with both Java nad Haskell backend.

I agree with the principle of keeping the lemmas for Java and Haskell in sync as much as possible.

Reply:
@denis-bogdanas Sorry I missed the newly added tests in the lemmas-spec.k. But my original question was the motivation of the change from:
```
    rule bool2Word( B:Bool ) ==Int 1 => B         [simplification]
    rule bool2Word( B:Bool ) ==Int 0 => notBool B [simplification]
```
to:
```
    rule bool2Word( B:Bool ) ==Int I => B ==K word2Bool(I)    [simplification, concrete(I)]
```
Did you find an actual case where `I` is other than 0 and 1, or is this just a refactoring change?  If the former, I want to know when it happens.

Yes, the new lemmas don't seem to affect the performance much, but I wanted to double check if you also thought about it.

Reply:
There are no such cases and they are not possible of course. The change was to (1) minimize number of lemmas in Haskell because they have symmetric equivalent so duplication gets multiplied, and (2) keep Java in sync with Haskell.